### ISSUE TITLE
Error: Unknown 'tableOf' option

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/prototype#210]
# What happened
I got the following error message:
<img width="927" alt="afbeelding" src="https://github.com/user-attachments/assets/59844897-e5f3-490d-a951-bd3da268925e">

# Analysis
This feels like I should not get error messages from a rabbit hole as deep as this. It should have been caught earlier. However, let's first analyze the error before jumping to conclusions.

This is the code that calls the transitive closure:
```Ampersand

ENFORCE connect[Connectie*Connectie] >: (van;naar~\/naar;van~) /\ berichttype;berichttype~

RELATION connectCopy[Connectie*Connectie] -- hiermee detecteren we een verandering in de populatie van `connect`
RELATION connectStar[Connectie*Connectie] -- de transitieve afsluiting van `connect`.
ROLE ExecEngine MAINTAINS "Compute `connectStar` after the population of `connect` has changed"
RULE "Compute `connectStar` after the population of `connect` has changed": connectCopy = connect
VIOLATION (TXT "{EX} TransitiveClosure;connect;Connectie;connectCopy;connectStar") -- This updates relations `connectCopy` and `connectStar`.  --}
```
I checked the code for Transitive closure in the prototype framework, but that looked unchanged for 7 years.

The error message 'Unknown 'tableOf' option for relation' occurs in the following code in `MysqlDB.php`, which is meant to delete all pairs in a relation:
```PHP
    public function emptyRelation(Relation $relation): void
    {
        $relationTable = $relation->getMysqlTable();

        switch ($relationTable->inTableOf()) {
            case null: // If n-n table, remove all rows
                $this->execute("DELETE FROM \"{$relationTable->getName()}\"");
                break;
            case 'src': // If in table of src concept, set tgt col to null
                $this->execute("UPDATE \"{$relationTable->getName()}\" SET \"{$relationTable->tgtCol()->getName()}\" = NULL");
                break;
            case 'tgt': // If in table of tgt concept, set src col to null
                $this->execute("UPDATE \"{$relationTable->getName()}\" SET \"{$relationTable->srcCol()->getName()}\" = NULL");
                break;
            default:
                throw new FatalException("Unknown 'tableOf' option for relation '{$relation}'");
        }
    }
```

I found no clues in the log file. 
```
prototype  | [2024-11-17 07:04:36] EXECENGINE.INFO: ExecEngine fixing 138 violations for rule 'Compute `connectStar` after the population of `connect` has changed' [] {"request_id":"d2aa9e8e27"}
prototype  | [2024-11-17 07:04:36] EXECENGINE.INFO: Fixing violation 1/138: ((LMS_0002[Connectie],LMS_0013[Connectie])) [] {"request_id":"d2aa9e8e27"}
prototype  | [2024-11-17 07:04:36] EXECENGINE.INFO: TransitiveClosure(connect,Connectie,connectCopy,connectStar) [] {"request_id":"d2aa9e8e27"}
prototype  | [2024-11-17 07:04:36] CORE.DEBUG: Deleting all links in relation connectCopy[Connectie*Connectie] [] {"request_id":"d2aa9e8e27"}
prototype  | [2024-11-17 07:04:36] API.ERROR: <h4>Error/Exception</h4><div><strong>Type:</strong> Ampersand\Exception\InvalidExecEngineCallException</div><div><strong>Message:</strong> Invalid exec engine function call as defined in /usr/local/project/Communicatie.adl:39:1. Function call &#039;TransitiveClosure(connect,Connectie,connectCopy,connectStar)&#039;. Unknown &#039;tableOf&#039; option for relation &#039;connectCopy[Connectie*Connectie]&#039;</div><div><strong>File:</strong> /var/www/src/Ampersand/Rule/ExecEngine.php</div><div><strong>Line:</strong> 244</div><div><strong>Trace:</strong><pre>#0 /var/www/src/Ampersand/Rule/ExecEngine.php(179): Ampersand\Rule\ExecEngine-&gt;fixViolation(Object(Ampersand\Rule\Violation)) #1 /var/www/src/Ampersand/Transaction.php(173): Ampersand\Rule\ExecEngine-&gt;checkFixRules(Array) #2 /var/www/src/Ampersand/Controller/PopulationController.php(56): Ampersand\Transaction-&gt;runExecEngine() #3 [internal function]: Ampersand\Controller\PopulationController-&gt;importPopulationFromUpload(Object(Slim\Http\Request), Object(Slim\Http\Response), Array) #4 /var/www/lib/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(40): call_user_func(Array, Object(Slim\Http\Request), Object(Slim\Http\Response), Array) #5 /var/www/lib/slim/slim/Slim/Route.php(281): Slim\Handlers\Strategies\RequestResponse-&gt;__invoke(Array, Object(Slim\Http\Request), Object(Slim\Http\Response), Array) #6 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(117): Slim\Route-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response)) #7 /var/www/lib/slim/slim/Slim/Route.php(268): Slim\Route-&gt;callMiddlewareStack(Object(Slim\Http\Request), Object(Slim\Http\Response)) #8 /var/www/lib/slim/slim/Slim/App.php(503): Slim\Route-&gt;run(Object(Slim\Http\Request), Object(Slim\Http\Response)) #9 /var/www/src/Ampersand/API/Middleware/LogPerformanceMiddleware.php(26): Slim\App-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response)) #10 [internal function]: Ampersand\API\Middleware\LogPerformanceMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Slim\App)) #11 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\LogPerformanceMiddleware), Array) #12 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Slim\App)) #13 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Slim\App)) #14 /var/www/src/Ampersand/API/Middleware/InitAmpersandAppMiddleware.php(54): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #15 [internal function]: Ampersand\API\Middleware\InitAmpersandAppMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #16 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\InitAmpersandAppMiddleware), Array) #17 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #18 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #19 /var/www/src/Ampersand/API/Middleware/PostMaxSizeMiddleware.php(27): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #20 [internal function]: Ampersand\API\Middleware\PostMaxSizeMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #21 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\PostMaxSizeMiddleware), Array) #22 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #23 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #24 /var/www/src/Ampersand/API/Middleware/JsonRequestParserMiddleware.php(22): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #25 [internal function]: Ampersand\API\Middleware\JsonRequestParserMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #26 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\JsonRequestParserMiddleware), Array) #27 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #28 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #29 /var/www/src/Ampersand/API/Middleware/LogPerformanceMiddleware.php(26): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #30 [internal function]: Ampersand\API\Middleware\LogPerformanceMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #31 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\LogPerformanceMiddleware), Array) #32 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #33 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #34 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(117): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #35 /var/www/lib/slim/slim/Slim/App.php(392): Slim\App-&gt;callMiddlewareStack(Object(Slim\Http\Request), Object(Slim\Http\Response)) #36 /var/www/lib/slim/slim/Slim/App.php(297): Slim\App-&gt;process(Object(Slim\Http\Request), Object(Slim\Http\Response)) #37 /var/www/bootstrap/framework.php(199): Slim\App-&gt;run() #38 /var/www/public/api/v1/index.php(3): require_once(&#039;/var/www/bootst...&#039;) #39 {main}</pre><h4>Error/Exception</h4><div><strong>Type:</strong> Ampersand\Exception\FatalException</div><div><strong>Message:</strong> Unknown &#039;tableOf&#039; option for relation &#039;connectCopy[Connectie*Connectie]&#039;</div><div><strong>File:</strong> /var/www/src/Ampersand/Plugs/MysqlDB/MysqlDB.php</div><div><strong>Line:</strong> 725</div><div><strong>Trace:</strong><pre>#0 /var/www/src/Ampersand/Core/Relation.php(302): Ampersand\Plugs\MysqlDB\MysqlDB-&gt;emptyRelation(Object(Ampersand\Core\Relation)) #1 /var/www/bootstrap/files/execEngineWarshall.php(65): Ampersand\Core\Relation-&gt;empty() #2 [internal function]: Ampersand\Rule\ExecEngine-&gt;{closure}(&#039;connect&#039;, &#039;Connectie&#039;, &#039;connectCopy&#039;, &#039;connectStar&#039;) #3 /var/www/src/Ampersand/Rule/ExecEngine.php(242): Closure-&gt;call(Object(Ampersand\Rule\ExecEngine), &#039;connect&#039;, &#039;Connectie&#039;, &#039;connectCopy&#039;, &#039;connectStar&#039;) #4 /var/www/src/Ampersand/Rule/ExecEngine.php(179): Ampersand\Rule\ExecEngine-&gt;fixViolation(Object(Ampersand\Rule\Violation)) #5 /var/www/src/Ampersand/Transaction.php(173): Ampersand\Rule\ExecEngine-&gt;checkFixRules(Array) #6 /var/www/src/Ampersand/Controller/PopulationController.php(56): Ampersand\Transaction-&gt;runExecEngine() #7 [internal function]: Ampersand\Controller\PopulationController-&gt;importPopulationFromUpload(Object(Slim\Http\Request), Object(Slim\Http\Response), Array) #8 /var/www/lib/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(40): call_user_func(Array, Object(Slim\Http\Request), Object(Slim\Http\Response), Array) #9 /var/www/lib/slim/slim/Slim/Route.php(281): Slim\Handlers\Strategies\RequestResponse-&gt;__invoke(Array, Object(Slim\Http\Request), Object(Slim\Http\Response), Array) #10 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(117): Slim\Route-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response)) #11 /var/www/lib/slim/slim/Slim/Route.php(268): Slim\Route-&gt;callMiddlewareStack(Object(Slim\Http\Request), Object(Slim\Http\Response)) #12 /var/www/lib/slim/slim/Slim/App.php(503): Slim\Route-&gt;run(Object(Slim\Http\Request), Object(Slim\Http\Response)) #13 /var/www/src/Ampersand/API/Middleware/LogPerformanceMiddleware.php(26): Slim\App-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response)) #14 [internal function]: Ampersand\API\Middleware\LogPerformanceMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Slim\App)) #15 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\LogPerformanceMiddleware), Array) #16 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Slim\App)) #17 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Slim\App)) #18 /var/www/src/Ampersand/API/Middleware/InitAmpersandAppMiddleware.php(54): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #19 [internal function]: Ampersand\API\Middleware\InitAmpersandAppMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #20 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\InitAmpersandAppMiddleware), Array) #21 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #22 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #23 /var/www/src/Ampersand/API/Middleware/PostMaxSizeMiddleware.php(27): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #24 [internal function]: Ampersand\API\Middleware\PostMaxSizeMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #25 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\PostMaxSizeMiddleware), Array) #26 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #27 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #28 /var/www/src/Ampersand/API/Middleware/JsonRequestParserMiddleware.php(22): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #29 [internal function]: Ampersand\API\Middleware\JsonRequestParserMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #30 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\JsonRequestParserMiddleware), Array) #31 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #32 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #33 /var/www/src/Ampersand/API/Middleware/LogPerformanceMiddleware.php(26): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #34 [internal function]: Ampersand\API\Middleware\LogPerformanceMiddleware-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #35 /var/www/lib/slim/slim/Slim/DeferredCallable.php(57): call_user_func_array(Object(Ampersand\API\Middleware\LogPerformanceMiddleware), Array) #36 [internal function]: Slim\DeferredCallable-&gt;__invoke(Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #37 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(70): call_user_func(Object(Slim\DeferredCallable), Object(Slim\Http\Request), Object(Slim\Http\Response), Object(Closure)) #38 /var/www/lib/slim/slim/Slim/MiddlewareAwareTrait.php(117): Slim\App-&gt;Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response)) #39 /var/www/lib/slim/slim/Slim/App.php(392): Slim\App-&gt;callMiddlewareStack(Object(Slim\Http\Request), Object(Slim\Http\Response)) #40 /var/www/lib/slim/slim/Slim/App.php(297): Slim\App-&gt;process(Object(Slim\Http\Request), Object(Slim\Http\Response)) #41 /var/www/bootstrap/framework.php(199): Slim\App-&gt;run() #42 /var/www/public/api/v1/index.php(3): require_once(&#039;/var/www/bootst...&#039;) #43 {main}</pre> [] {"request_id":"d2aa9e8e27"}
```
The top of the stack looks like this:
```
Error/Exception
Type: Ampersand\Exception\InvalidExecEngineCallException
Message: Invalid exec engine function call as defined in /usr/local/project/Communicatie.adl:39:1. Function call 'TransitiveClosure(connect,Connectie,connectCopy,connectStar)'. Unknown 'tableOf' option for relation 'connectCopy[Connectie*Connectie]'
File: /var/www/src/Ampersand/Rule/ExecEngine.php
Line: 244
Trace:

#0 /var/www/src/Ampersand/Rule/ExecEngine.php(179): Ampersand\Rule\ExecEngine->fixViolation(Object(Ampersand\Rule\Violation))
#1 /var/www/src/Ampersand/Transaction.php(173): Ampersand\Rule\ExecEngine->checkFixRules(Array)
#2 /var/www/src/Ampersand/Controller/PopulationController.php(56): Ampersand\Transaction->runExecEngine()
#3 [internal function]: Ampersand\Controller\PopulationController->importPopulationFromUpload(Object(Slim\Http\Request), Object(Slim\Http\Response), Array)
#4 /var/www/lib/slim/slim/Slim/Handlers/Strategies/RequestResponse.php(40): call_user_func(Array, Object(Slim\Http\Request), Object(Slim\Http\Response), Array)
```
I find it difficult to make this problem reproducible, but hopefully, we'll get there by reasoning alone.

# Version of ampersand that was used

I used the prototype framework v1.17.0. This implies Ampersand v4.7.

**Comment by hanjoosten** [comment]
Note that the use of TransitiveClosure will become deprecated. It is advised to use the cleaner (Kleener ?) * and + expression in ampersand 5, which will hopefully be available soon

**Comment by hanjoosten** [comment]
Maybe it is a good idea to extend the text thrown by `FatalException` with the outcome of `$relationTable->inTableOf()`

**Comment by stefjoosten** [comment]
I have extended the text, as suggested by Han. This is the result:
<img width="933" alt="afbeelding" src="https://github.com/user-attachments/assets/513c0539-6319-447e-86c8-feeefd5701e2">

**Comment by stefjoosten** [comment]
# Solution
I have no idea about the root cause of this problem. However, I can cosmetically cover this bug by treating the case `binary` as though it were null. This is a wild guess, but it gets rid of the bug. If this guess turns out wrong, we probably have a bigger problem because in that case, I have hidden an existing bug behind a new one. So, @Michiel-s, can you verify this solution? 
```php
    public function emptyRelation(Relation $relation): void
    {
        $relationTable = $relation->getMysqlTable();
        $tableOf = $relationTable->inTableOf()->value;

        switch ($tableOf) {
            case null: // If n-n table, remove all rows
                $this->execute("DELETE FROM \"{$relationTable->getName()}\"");
                break;
            case 'src': // If in table of src concept, set tgt col to null
                $this->execute("UPDATE \"{$relationTable->getName()}\" SET \"{$relationTable->tgtCol()->getName()}\" = NULL");
                break;
            case 'tgt': // If in table of tgt concept, set src col to null
                $this->execute("UPDATE \"{$relationTable->getName()}\" SET \"{$relationTable->srcCol()->getName()}\" = NULL");
                break;
            case 'binary': // If n-n table, remove all rows
                $this->execute("DELETE FROM \"{$relationTable->getName()}\"");
                break;
            default:
                throw new FatalException("Unknown 'tableOf' option '{$tableOf}' for relation '{$relation}'");
        }
    }
```
This intervention helps me to continue my work, so for now I'm deploying this locally.

**Comment by Michiel-s** [comment]
Please note that the fix is applied in main branch by #214, but also applied on v1 and released in [v1.18.1](https://github.com/AmpersandTarski/prototype/releases/tag/v1.18.1) and applied on v2 branch.
